File: apps/backend/src/modules/trips/dto/create-trip.dto.ts
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import {
  IsString,
  IsOptional,
  IsEnum,
  IsNumber,
  IsInt,
  IsBoolean,
  IsDateString,
  ValidateNested,
  Min,
} from 'class-validator';
import { TripStatus } from '@prisma/client';

export class TripLocationDto {
  @ApiPropertyOptional({ example: 'Magazzino Nord', description: 'Nome azienda/luogo' })
  @IsOptional()
  @IsString()
  companyName?: string;

  @ApiProperty({ example: 'Via Industria 45', description: 'Indirizzo' })
  @IsString()
  address: string;

  @ApiProperty({ example: 'Milano', description: 'Città' })
  @IsString()
  city: string;

  @ApiProperty({ example: 'MI', description: 'Provincia' })
  @IsString()
  province: string;

  @ApiProperty({ example: '20100', description: 'CAP' })
  @IsString()
  postalCode: string;

  @ApiPropertyOptional({ example: 'Italia', default: 'Italia', description: 'Paese' })
  @IsOptional()
  @IsString()
  country?: string;

  @ApiPropertyOptional({ example: 45.4642, description: 'Latitudine' })
  @IsOptional()
  @IsNumber()
  lat?: number;

  @ApiPropertyOptional({ example: 9.19, description: 'Longitudine' })
  @IsOptional()
  @IsNumber()
  lng?: number;
}

export class CargoInfoDto {
  @ApiProperty({ example: 'Materiale elettronico', description: 'Descrizione carico' })
  @IsString()
  description: string;

  @ApiProperty({ example: 5000, description: 'Peso in kg' })
  @IsNumber()
  @Min(0)
  weight: number;

  @ApiPropertyOptional({ example: 25, description: 'Volume in m³' })
  @IsOptional()
  @IsNumber()
  volume?: number;

  @ApiPropertyOptional({ example: 10, description: 'Numero colli' })
  @IsOptional()
  @IsInt()
  packages?: number;

  @ApiPropertyOptional({ default: false, description: 'Merce ADR (pericolosa)' })
  @IsOptional()
  @IsBoolean()
  isADR?: boolean;

  @ApiPropertyOptional({ example: -18, description: 'Temperatura richiesta (°C)' })
  @IsOptional()
  @IsNumber()
  temperature?: number;
}

export class CreateTripDto {
  @ApiProperty({ example: 'VG-2024-001', description: 'Numero viaggio' })
  @IsString()
  tripNumber: string;

  @ApiProperty({ description: 'ID veicolo' })
  @IsString()
  vehicleId: string;

  @ApiProperty({ description: 'ID autista' })
  @IsString()
  driverId: string;

  @ApiProperty({ description: 'ID cliente' })
  @IsString()
  clientId: string;

  @ApiProperty({ type: TripLocationDto, description: 'Luogo di partenza' })
  @ValidateNested()
  @Type(() => TripLocationDto)
  origin: TripLocationDto;

  @ApiProperty({ type: TripLocationDto, description: 'Luogo di destinazione' })
  @ValidateNested()
  @Type(() => TripLocationDto)
  destination: TripLocationDto;

  @ApiProperty({ type: CargoInfoDto, description: 'Informazioni carico' })
  @ValidateNested()
  @Type(() => CargoInfoDto)
  cargo: CargoInfoDto;

  @ApiPropertyOptional({
    enum: TripStatus,
    default: 'planned',
    description: 'Stato viaggio',
  })
  @IsOptional()
  @IsEnum(TripStatus)
  status?: TripStatus;

  @ApiProperty({ example: '2024-12-01T08:00:00Z', description: 'Partenza pianificata' })
  @IsDateString()
  plannedDeparture: string;

  @ApiPropertyOptional({ description: 'Partenza effettiva' })
  @IsOptional()
  @IsDateString()
  actualDeparture?: string;

  @ApiProperty({ example: '2024-12-01T18:00:00Z', description: 'Arrivo pianificato' })
  @IsDateString()
  plannedArrival: string;

  @ApiPropertyOptional({ description: 'Arrivo effettivo' })
  @IsOptional()
  @IsDateString()
  actualArrival?: string;

  @ApiProperty({ example: 500, description: 'Km pianificati' })
  @IsInt()
  @Min(0)
  kmPlanned: number;

  @ApiPropertyOptional({ description: 'Km effettivi' })
  @IsOptional()
  @IsInt()
  kmActual?: number;

  @ApiProperty({ example: 850.50, description: 'Prezzo trasporto (€)' })
  @IsNumber()
  @Min(0)
  price: number;

  @ApiPropertyOptional({ description: 'Note aggiuntive' })
  @IsOptional()
  @IsString()
  notes?: string;
}

File: apps/backend/src/main/java/com/truckfleet/dto/trip/CreateTripDto.java
package com.truckfleet.dto.trip;

import com.truckfleet.entity.enums.TripStatus;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import lombok.Data;

import java.math.BigDecimal;

@Data
@Schema(description = "DTO per la creazione di un viaggio")
public class CreateTripDto {

    @NotBlank(message = "Il numero viaggio e obbligatorio")
    @Schema(example = "VGG-2024-001", description = "Numero viaggio")
    private String tripNumber;

    @NotBlank(message = "L'ID veicolo e obbligatorio")
    @Schema(description = "ID del veicolo")
    private String vehicleId;

    @NotBlank(message = "L'ID autista e obbligatorio")
    @Schema(description = "ID dell'autista")
    private String driverId;

    @NotBlank(message = "L'ID cliente e obbligatorio")
    @Schema(description = "ID del cliente")
    private String clientId;

    @NotNull(message = "L'origine e obbligatoria")
    @Valid
    @Schema(description = "Luogo di partenza")
    private TripLocationDto origin;

    @NotNull(message = "La destinazione e obbligatoria")
    @Valid
    @Schema(description = "Luogo di arrivo")
    private TripLocationDto destination;

    @NotNull(message = "Le informazioni sul carico sono obbligatorie")
    @Valid
    @Schema(description = "Informazioni sul carico")
    private CargoInfoDto cargo;

    @Schema(description = "Stato del viaggio")
    private TripStatus status;

    @NotBlank(message = "La partenza pianificata e obbligatoria")
    @Schema(example = "2024-01-15T08:00:00", description = "Data/ora partenza pianificata")
    private String plannedDeparture;

    @Schema(example = "2024-01-15T08:30:00", description = "Data/ora partenza effettiva")
    private String actualDeparture;

    @NotBlank(message = "L'arrivo pianificato e obbligatorio")
    @Schema(example = "2024-01-15T18:00:00", description = "Data/ora arrivo pianificato")
    private String plannedArrival;

    @Schema(example = "2024-01-15T17:30:00", description = "Data/ora arrivo effettivo")
    private String actualArrival;

    @NotNull(message = "I km pianificati sono obbligatori")
    @Min(value = 0, message = "I km pianificati non possono essere negativi")
    @Schema(example = "450", description = "Km pianificati")
    private Integer kmPlanned;

    @Schema(example = "445", description = "Km effettivi")
    private Integer kmActual;

    @NotNull(message = "Il prezzo e obbligatorio")
    @Min(value = 0, message = "Il prezzo non puo essere negativo")
    @Schema(example = "1500.00", description = "Prezzo del trasporto")
    private BigDecimal price;

    @Schema(description = "Note")
    private String notes;
}
